### PR TITLE
feat: implement OCI image analysis

### PR DESCRIPTION
**Pull request by Strum355**
## Description

> Describe what you did and why.

**Related issues (if any):** 

- fixes: #issue_number_goes_here

## Checklist

- [x] I have followed this repository's contributing guidelines.
- [x] I will adhere to the project's code of conduct.

## Additional information

> Anything else?

### CODE REVIEW

File: src/oci_image/images.js [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+*
+* Example of valid names:
+*
+* - consol/tomcat-8.0
+* - consol/tomcat-8.0:8.0.9
+* - docker.consol.de:5000/tomcat-8.0
+* - docker.consol.de:5000/jolokia/tomcat-8.0:8.0.9
+*/
+export class Image {
+	static nameComponentRegexp = '[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?';
```
Comment (ruromero) [inline src/oci_image/images.js:20]: It's not a good practice to name variables/constants after the type they have. Also if they're supposed to be immutable you can use snake case for all. I suggest the following.

```suggestion
	static NAME_COMP_PATTERN = '[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?';
	static DOMAIN_COMP_PATTERN = '(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])';
	static NAME_COMP_REGEX = new RegExp(this.NAME_COMP_PATTERN);
	static IMAGE_NAME_REGEX = new RegExp(this.NAME_COMP_PATTERN + '(?:(?:/' + this.NAME_COMP_PATTERN + ')+)?');
	static DOMAIN_REGEX = new RegExp('^' + this.DOMAIN_COMP_PATTERN + '(?:\\.' + this.DOMAIN_COMP_PATTERN + ')*(?::[0-9]+)?$');
	static TAG_REGEX = new RegExp('^[\\w][\\w.-]{0,127}$');
	static DIGEST_REGEX = new RegExp('^sha256:[a-z0-9]{32,}$');
```

File: src/oci_image/images.js [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	 * @see https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#oci
+	 */
+	getPackageURL() {
+		/** @type {Object.<string, string>} */
+		const qualifiers = {};
+		const repositoryUrl = this.image.getNameWithoutTag();
+		const simpleName = this.image.getSimpleName();
+
+		if (repositoryUrl != null && repositoryUrl.toLowerCase() !== simpleName.toLowerCase()) {
+			qualifiers[ImageRef.REPOSITORY_QUALIFIER, repositoryUrl.toLowerCase()];
```
Comment (ruromero) [inline src/oci_image/images.js:291]: I think you are trying to assign the value here:
```suggestion
			qualifiers[ImageRef.REPOSITORY_QUALIFIER] = repositoryUrl.toLowerCase();
```

File: src/oci_image/images.js [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+		}
+
+		if (this.platform != null) {
+			qualifiers[ImageRef.ARCH_QUALIFIER, this.platform.architecture.toLowerCase()];
+			qualifiers[ImageRef.OS_QUALIFIER, this.platform.os.toLowerCase()];
+			if (this.platform.variant != null) {
+				qualifiers[ImageRef.VARIANT_QUALIFIER, this.platform.variant.toLowerCase()];
```
Comment (ruromero) [inline src/oci_image/images.js:298]: Same
```suggestion
				qualifiers[ImageRef.VARIANT_QUALIFIER] = this.platform.variant.toLowerCase();
```

File: src/oci_image/images.js [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+			}
+		}
+
+		const tag = this.image.tag;
+		if (tag != null) {
+			qualifiers[ImageRef.TAG_QUALIFIER, tag];
```
Comment (ruromero) [inline src/oci_image/images.js:304]:
```suggestion
			qualifiers[ImageRef.TAG_QUALIFIER] = tag;
```

File: src/oci_image/images.js (hunk 3 above)
Comment (ruromero) [inline src/oci_image/images.js:298]:
```suggestion
				qualifiers[ImageRef.VARIANT_QUALIFIER] = this.platform.variant.toLowerCase();
```

File: src/analysis.js [hunk 5]
```diff
@@ -133,6 +134,54 @@ async function requestComponent(provider, manifest, url, opts = {}) {
...
+		method: 'POST',
+		headers: {
+			'Accept': html ? 'text/html' : 'application/json',
+			'Content-Type': 'application/vnd.cyclonedx+json',
+			...getTokenHeaders(opts)
+		},
+		body: JSON.stringify(imageSboms),
+	})
+
+	console.error(JSON.stringify(imageSboms, '', '\t'))
```
Comment (ruromero) [inline src/analysis.js:161]: Remove this line
```suggestion
```

File: test/providers/oci_images.test.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { expect } from "chai";
+import fs from 'fs';
+import sinon from 'sinon'
+import { generateImageSBOM, parseImageRef } from '../../src/oci_image/utils.js';
+
+let clock;
+suite('testing the OCI image data provider', () => {
```
Comment (ruromero) [inline test/providers/oci_images.test.js:7]: It would be good if you added some more tests. Specially for the corner cases like invalid platforms or image strings.
Also you could implement specific unit tests to validate the individual functionality. 
Given our current situation, consider this optional.

**Review by ruromero — CHANGES_REQUESTED** [review]
This one wasn't trivial. Good job.

**Review by ruromero — APPROVED** [review]
